accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,706 @@
+See the existing [ILGenerator.EmitCalli()](https://docs.microsoft.com/dotnet/api/system.reflection.emit.ilgenerator.emitcalli#system-reflection-emit-ilgenerator-emitcalli) for reference.
+```diff
+namespace System.Reflection.Emit
+{
+    public class ILGenerator
+    {
++       public virtual void EmitCalli(
+            // Must be OpCodes.Calli which matches existing EmitCalli convention
+            // Not sure why this is redundant (for future Calli opcodes?)
++           System.Reflection.Emit.OpCodes opcode,
++           Type[] callingConventions,
++           Type? returnType,
++           Type[]? parameterTypes);
+
++       public virtual void EmitCalli(
++           System.Reflection.Emit.OpCodes opcode,
++           Type functionPointerType)
+    }
+}
+```
+
+# Breaking changes
+The breaking change is: **`typeof(delegate*<...>)` and `Type.GetType()` will return a `Type` instance instead of `IntPtr`**
+
+The fallout from this is that for any general-purpose reflection code that wants to support function pointers will likely need to call `Type.IsFunctionPointer` if there is special treatment for function pointers. 
+
+# Potential future features
+Various features not planned for 8.0; somewhat discussed earlier.
+
+## Support `Type.FullName` and `Type.AssemblyQualifiedName`
+Including `Type.GetType()` and similar APIs to parse that string to construct a function pointer type dynamically.
+
+## `System.Reflection.FunctionPointer` for `System.Object`-based invoke
+See the existing [System.Reflection.Pointer](https://docs.microsoft.com/dotnet/api/system.reflection.pointer) for reference. This would allow passing strongly-typed, validated function pointers when using `System.Object`-based invoke. Naming of Box\Unbox vs. constructors etc. is TBD.
+
+ This is a value type, not a reference type like `Pointer`, since newer proposed reflection APIs will not force boxing to occur.
+
+ Like `Pointer` there is no property to obtain the type since it is used in a temporary manner to manually marshal.
+
+```diff
+namespace System.Reflection
+{
++   public struct FunctionPointer
++   {
++       public static FunctionPointer Box(void* ptr, Type type);
++       public static void* Unbox (FunctionPointer ptr);
++   }
+}
+```
+
+## Support Type.GetMethod()
+If a user tries to retrieve a Method through [`Type.GetMethod()`](https://learn.microsoft.com/dotnet/api/system.type.getmethod) that has function pointer arguments specified by the `Type[]` passed into that method, the current design throws `NotSupportedException` since we don't have a good way to specify the custom modifiers for each parameter in the API (see below -- requires "MethodSignature").
+
+The workaround is to have the caller manually loop through the `MethodInfo`s returned from `Type.GetMethods()` and perform the filtering there.
+
+```diff
+    public abstract class Type
+    {
+        // This is used with 'Type.GetMethod(..., Type[] types)' to find a method with parameters that
```

Comment:
Can `typeof(delegate* <void>)` be used with `Type.GetMethod` as well (for runtime types)? Is `ReferenceEquals(Type.MakeFunctionPointerType(typeof(void), Types.EmptyTypes, false) , typeof(delegate* <void>))` going to be `true`? 

I do not see how this API can work well with different type universes, but we do not need to figure this out now that it is in the potential future additions section.

Reply:
> Can typeof(delegate*<void>) be used with Type.GetMethod as well (for runtime types)?

Good question. IMO it does make sense to support runtime types like this (in addition to those from MakeFunctionPointerType). I'll add a note to the design for this.

> Is ReferenceEquals(Type.MakeFunctionPointerType(typeof(void), Types.EmptyTypes, false) , typeof(delegate* <void>)) going to be true?

No, unless we want to do something more here than just supporting `GetMethod()` lookups. The method comparison logic in `GetMethod()` would match up parameter types\modifiers, so don't need to require `==`, `Equals()` or a `ReferenceEquals()` match. Similar to`Type.MakeGenericSignatureType()`, a type returned from `MakeFunctionPointerType()` will derive from [`SignatureType`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Reflection/SignatureType.cs) and isn't expected to be used outside of GetMethod() lookups.

Reply:
> > Is ReferenceEquals(Type.MakeFunctionPointerType(typeof(void), Types.EmptyTypes, false) , typeof(delegate* )) going to be true?

> No, unless we want to do something more here than just supporting GetMethod() lookups.

I think we should do more here if/once we get to this.

All existing `Make*` reflection APIs return the type that you would get by other means (e.g. by inspecting method or field signatures). It would be confusing for `MakeFunctionPointer` to return a function pointer type that is not full-fledged function pointer type.

`MakeGenericSignatureType` is special. It is not possible to get the type returned by `MakeGenericSignatureType` by other means and that's why it does not need to unify with anything.

Reply:
> I think we should do more here if/once we get to this.

Sure - I'll update the design to suggest the full approach.